velox/substrait/SubstraitExtensionCollector.h

```diff
@@ -0,0 +1,118 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <optional>
+#include "velox/core/Expressions.h"
+#include "velox/core/PlanNode.h"
+#include "velox/substrait/VeloxSubstraitSignature.h"
+#include "velox/substrait/proto/substrait/algebra.pb.h"
+#include "velox/substrait/proto/substrait/plan.pb.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::substrait {
+
+struct ExtensionFunctionId {
+  /// Substrait extension YAML file uri.
+  std::string uri;
+
+  /// Substrait signature used in the function extension declaration is a
+  /// combination of the name of the function along with a list of input
+  /// argument types.The format is as follows : <function
+  /// name>:<short_arg_type0>_<short_arg_type1>_..._<short_arg_typeN> for more
+  /// detail information about the argument type please refer to link
+  /// https://substrait.io/extensions/#function-signature-compound-names.
+  std::string signature;
+
+  bool operator==(const ExtensionFunctionId& other) const {
+    return (uri == other.uri && signature == other.signature);
+  }
+};
+
+/// Assigns unique IDs to function signatures using ExtensionFunctionId.
+class SubstraitExtensionCollector {
+ public:
+  SubstraitExtensionCollector();
+
+  /// Given a scalar function name and argument types, return the functionId
+  /// using ExtensionFunctionId.
+  int getReferenceNumber(
+      const std::string& functionName,
+      const std::vector<TypePtr>& arguments);
+
+  /// Given an aggregate function name and argument types and aggregation Step,
+  /// return the functionId using ExtensionFunctionId.
+  int getReferenceNumber(
+      const std::string& functionName,
+      const std::vector<TypePtr>& arguments,
+      core::AggregationNode::Step aggregationStep);
+
+  /// Add extension functions to Substrait plan.
+  void addExtensionsToPlan(::substrait::Plan* plan) const;
+
+ private:
+  /// A bi-direction hash map to keep the relation between reference number and
+  /// either function or type signature.
+  /// @T ExtensionFunctionId
+  template <class T>
+  class BiDirectionHashMap {
+   public:
+    /// If the forwardMap_ does not contain the key, then the key and value will
+    /// be written, otherwise the value will be overwritten, If the reverseMap_
+    /// does not contain the value, then the value and key will be written,
+    /// otherwise the key will be overwritten.
+    void put(const int& key, const T& value);
+
+    std::map<int, T>& forwardMap() {
```

Comment:
These methods should be const and should return const &. Also, consider replacing these with lookups by key and value, i.e.

```
std::optional<int> findKey(value) const;
std::optional<T> findValue(key) const;
```

Reply:
change to const is ok, but if replace with lookup methods that won't work for addExtensionsToPlan that need to iterate over the forwardMap

Reply:
I see. I guess we still can make this method const and have it return const &. No need to provide write access to forwardMap_, right?

Reply:
Yes, updated